driver-core/src/main/com/mongodb/internal/async/AsyncSupplier.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.async;
+
+import java.util.function.Predicate;
+
+import static com.mongodb.internal.async.AsyncRunnable.CallbackThrew;
+
+/**
+ * See AsyncRunnableTest for usage
+ */
+public interface AsyncSupplier<T> {
+
+    void supplyUnsafe(SingleResultCallback<T> callback);
+
+    /**
+     * Must be invoked at end of async chain
+     * @param callback the callback provided by the method the chain is used in
+     */
+    default void finish(final SingleResultCallback<T> callback) {
+        try {
+            this.supplyUnsafe((v, e) -> {
+                try {
+                    callback.onResult(v, e);
+                } catch (Throwable t) {
+                    throw new CallbackThrew("Unexpected Throwable thrown from callback: ", e);
+                }
+            });
+        } catch (CallbackThrew t) {
+            // ignore
+        } catch (Throwable t) {
+            callback.onResult(null, t);
+        }
```

Comment:
For onResultUnsafe, I am unsure that it is worth always wrapping SingleResultCallback in another call, in order to ensure that exceptions are wrapped in assertion errors. Doesn't the existing finish method already handle this in one place? (As an aside, we should not wrap assertion errors in an assertion error.)

For double-invocation of onResult, I agree, in cases of VM errors, and in cases (where the user violates the "method must return after an async call") OR (where both the preceding and ensuing callbacks violate the contract), in a certain way that conceals or transforms the CallbackThrew assertion error, the callback would be invoked twice, with the second invocation being on the error path. These factors combined, this seems like this will address a minor risk, but would also introduce state into the closure.

Reply:
> As an aside, we should not wrap assertion errors in an assertion error

I agree. I don't think I proposed doing this either in plain English or in the proposed code.

> For onResultUnsafe, I am unsure that it is worth always wrapping SingleResultCallback in another call, in order to ensure that exceptions are wrapped in assertion errors. Doesn't the existing finish method already handle this in one place?

If done in `SingleResultCallback`, no code ever needs to care about this. If done in `finish`

- it will have to be done in all `finish` implementations if there is more than one;
- and in all places that complete callbacks not via the API you introduce (yeah, we don't do that, but the proposal I made fixes that).

Reply:
I added tests for double-invocations.

Unaffected async code SHOULD be preserved, including for example null checks, which throw exceptions, so chained lambdas MAY throw exceptions, which are expected to be handled in exactly one place (the finish method, when it is invoked at the end of the chain).